MapboxNavigation/NavigationMapView.swift

```diff
@@ -427,36 +426,82 @@ open class NavigationMapView: MGLMapView {
+        
+        let waypointTest = waypoints(on: routes, closeTo: tapPoint) //are there waypoints near the tapped location?
+        if let selected = waypointTest?.first { //test passes
+            navigationMapDelegate?.navigationMapView?(self, didSelectWaypoint: selected)
+            return
+        } else if let routes = self.routes(closeTo: tapPoint) {
+            guard let selectedRoute = routes.first else { return }
+            select(route: selectedRoute)
+        }
+        
+    }
+    
+    //TODO: Change to point-based distance calculation
+    private func waypoints(on routes: [Route], closeTo point: CGPoint) -> [Waypoint]? {
+        let tapCoordinate = convert(point, toCoordinateFrom: self)
+        let waypointThreshold: CGFloat = 100 //FIXME: Class Constant
+        let multipointRoutes = routes.filter { $0.routeOptions.waypoints.count >= 3}
+        guard multipointRoutes.count > 0 else { return nil }
+        let waypoints = multipointRoutes.flatMap({$0.routeOptions.waypoints})
+        
+        //lets sort the array in order of closest to tap
+        let closest = waypoints.sorted { (left, right) -> Bool in
+            let leftDistance = left.coordinate.distance(to: tapCoordinate)
+            let rightDistance = right.coordinate.distance(to: tapCoordinate)
             return leftDistance < rightDistance
         }
         
-        if let closestRoute = closestRoute, let closestCoordinate = Polyline(closestRoute.coordinates!).closestCoordinate(to: tapCoordinate), let routeIndex = routes.index(where: { $0 ==  closestRoute}) {
-            let closestPoint = self.convert(closestCoordinate.coordinate, toPointTo: self)
-            let tapDistanceFromClosestRoute = closestPoint.distance(to: tapPoint)
-            if tapDistanceFromClosestRoute <= maximumTapDistanceToSelectAltRoute {
-                let selectedRoute = routes[routeIndex]
-                self.routes?.remove(at: routeIndex)
-                self.routes?.insert(selectedRoute, at: 0)
-                self.showRoutes(self.routes!)
-                navigationMapDelegate?.navigationMapView?(self, didTap: selectedRoute)
-            }
+        //lets filter to see which ones are under threshold
+        let candidates = closest.filter({
+            let coordinatePoint = self.convert($0.coordinate, toPointTo: self)
+            return coordinatePoint.distance(to: point) < waypointThreshold
+        })
+        
+        return candidates
+    }
+    
+    private func routes(closeTo point: CGPoint) -> [Route]? {
+        let tapCoordinate = convert(point, toCoordinateFrom: self)
+        
+        //do we have routes? If so, filter routes with at least 2 coordinates.
+        guard let routes = routes?.filter({ $0.coordinates?.count ?? 0 > 1 }) else { return nil }
+        
+        //Sort routes by closest distance to tap gesture.
+        let closest = routes.sorted { (left, right) -> Bool in
```

Comment:
How does this sort perform for long routes, such as cross-country routes?

As long as we can figure out which route polyline feature on the map corresponds to which Route object – perhaps by assigning each feature a unique identifier – we can query the map view for visible features near the tap point to determine the nearest route. Feature querying should be more performant, since it’s powered by an R-tree implemented in C++.

Reply:
Strongly agree, Will fix. Ideally it would be nice if the map told the implementer that a map feature was selected (much like we tell a implementer that an annotation has been selected), but that's a different battle, for a different war.

Reply:
> Ideally it would be nice if the map told the implementer that a map feature was selected (much like we tell an implementer that an annotation has been selected)

This would be a great feature to have in the map SDK, since it isn’t specific to navigation: mapbox/mapbox-gl-native#6515.